File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    left, right = 1, max(diffs)
    answer = right
    
    def can_finish(level):
        total = times[0] # 첫번째 퍼즐은 항상 한번에 풀 수 있음
        if total > limit:
            return False
        
        n = len(diffs)
        for i in range(1, n):
            k = max(0, diffs[i] - level)
            total += times[i] + (times[i-1] + times[i]) * k
            if total > limit:
                return False
        return True
        
    while left <= right:
        mid = (left + right) // 2
        if can_finish(mid):
            answer = mid
            right = mid - 1
        else:
            left = mid + 1
        
    return answer

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
import java.util.*;

class Solution {
    public int solution(int[] diffs, int[] times, long limit) {
        int left = 1, right = Arrays.stream(diffs).max().getAsInt();
        int answer = right;
        
        while (left <= right) {
            int mid = (left + right) / 2;
            if (canSolve(mid, diffs, times, limit)) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return answer;
    }
    
    private boolean canSolve(int level, int[] diffs, int[] times, long limit) {
        long total = times[0];
        if (total > limit) return false;
        
        for (int i = 1; i < diffs.length; i++) {
            int k = Math.max(0, diffs[i] - level);
            total += (long) times[i] + (long) (times[i-1] + times[i]) * k;
            if (total > limit) return false;
        }
        return true;
    }
}